### DISCUSSION TITLE
Is there any reasons to use service accounts instead of clinets when it comes to call API using the Client Credentials Flow?

### DISCUSSION THREAD

**Discussion by kimminw00** [keycloak/keycloak#26190] [Q&A]
> Clients are applications and services that can request authentication of a user.

From the above line, I think that clients are already thought of as service accounts.
But each OIDC client has a built-in service account.
Is there any reasons to use service accounts instead of clinets when it comes to call API using the Client Credentials Flow?

**Comment by darius-m** [comment]
I don't think clients are seen as users, but rather they are services that request authentication *on behalf* of users. I don't think clients themselves have API access to Keycloak, but rather they can access some endpoints or redirect users in order to get user information in the form of JWTs (for OIDC) or documents / assertions (for SAML) after the user is authenticated. However, a client cannot access the Keycloak API to, for example, check if a (random) user exists. Note that this is different from an authentication or authorization flow, which requires the user to enter their credentials, and the client can only access data related to the user who is authenticating.

**Reply by kimminw00** [reply]
> [4.4](https://datatracker.ietf.org/doc/html/rfc6749#section-4.4).  Client Credentials Grant 
The client can request an access token using only its client
 credentials (or other supported means of authentication) when the
 client is requesting access to the protected resources under its
 control, or those of another resource owner that have been previously
 arranged with the authorization server (the method of which is beyond
 the scope of this specification).

From the above lines, the client can request an access token using only its client credentials (or other supported means of authentication). Then, other services which has the issued access token can call APIs(e.g. webserver's APIs). 

> Client credentials grant
The Client Credentials Grant creates a token based on the metadata and permissions of a service account associated with the client instead of obtaining a token that works on behalf of an external user. Client Credentials Grants are used by REST clients. See the [Service Accounts](https://www.keycloak.org/docs/latest/server_admin/#_service_accounts) chapter for more information.

And keycloak chose to use service account as means of authentication. I'm asking why keycloak chose service accounts rather than clients even though access tokens can be issued from clients.

**Reply by darius-m** [reply]
I think the two are basically the same explanation, but I suppose some clarification for the first paragraph could be in order.

*From my understanding*, a client is an abstraction of the service that requires authentication. The client basically defines "who can request" authentication (a client ID), "how to validate" that the request is valid (redirect URIs, client secrets) and "what data" should be sent to the service (through mappers) after user authentication. There is normally no need for a client to be able to perform API interactions, so the client abstraction can be kept somewhat small.

If there is, for any reason, a need to be able to perform API requests, the service account is added so the service can authenticate to Keycloak and perform operations based on its own permissions, instead of those of a regular user.

Not sure whether the implementation details are covered in the OAuth2 / OIDC specifications, but having a separate abstraction in the form of a service account likely helps with decreasing duplicated functionality - most of what the service account can (or must) do overlaps with regular users - e.g., user authentication (and the associated bearer access token), permissions, access logs. For basically all intents and purposes, the service account is the same as a regular user, with notable differences being that there is no long-term session created after the account logs in, and the account can log in using the client ID and secret instead of going through the regular authentication flows.
